src/wasmtime/component.rs

```diff
@@ -107,7 +107,60 @@ impl WasmValue for component::Val {
     type Type = component::Type;
 
     fn ty(&self) -> Self::Type {
-        self.ty()
+        match self {
+            Self::Bool(_) => component::Type::Bool,
+            Self::S8(_) => component::Type::S8,
+            Self::U8(_) => component::Type::U8,
+            Self::S16(_) => component::Type::S16,
+            Self::U16(_) => component::Type::U16,
+            Self::S32(_) => component::Type::S32,
+            Self::U32(_) => component::Type::U32,
+            Self::S64(_) => component::Type::S64,
+            Self::U64(_) => component::Type::U64,
+            Self::Float32(_) => component::Type::Float32,
+            Self::Float64(_) => component::Type::Float64,
+            Self::Char(_) => component::Type::Char,
+            Self::String(_) => component::Type::String,
+            _ => unimplemented!(),
```

Comment:
I think it would be better to remove this method from the trait entirely rather than allow it to panic. Would that work with the changes in this PR?